internal/server/pool_test.go

```diff
@@ -53,7 +54,455 @@ func TestPoolSpike(t *testing.T) {
 	wg.Wait()
 
 	be.Equal(rt, 5, p.Stats().Idle)
-	be.NilErr(rt, p.Close())
+}
+
+func TestMaxOpenConnsUpdate(t *testing.T) {
+	t.Run("basic", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 2
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn2, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+
+		p.SetMaxOpenConns(2)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		// Ensure that extra connections are not added back.
+		be.Equal(t, 2, p.Stats().OpenConnections)
+
+		p.ReleaseConn(conn2)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
```

Comment:
Probably a dumb question, but I'll selfishly use this review for my own learning:
Why does releasing a connection keep the number of open connections the same?

Reply:
Yes, good one. The difference is between `OpenConnections` and `InUse` connections. What you are thinking of is `InUse`. `OpenConnections` is always the number of connections opened to the DB.

https://github.com/mattermost/perseus/blob/10f1c2254aa7941ee8f8ae1ae407be6af61b2bae/internal/server/pool.go#L686-L688

Reply:
Aaah, I see, I see. Thanks :)